design/open/reuse-contented-merge-metaranges/reuse-contented-merge-metaranges.md

```diff
@@ -0,0 +1,128 @@
+
+* A range that was the same in the base and the destination, but changed
+  only between base and original source will still generally be the same in
+  the base and destination.  And rewriting it from the new source yields the
+  same range -- it will be reused in the result.
+
+  **Result:** The range created in the first attempt will be recreated in
+  the second attempt, and will not need to be uploaded to the backing object
+  store.  (We might later decide to cache the merged range result in memory,
+  which will let us skip even recreating that range!)
+* A range that was the same in the base and the source, but changed only
+  between base and original destination, will still be unchanged to the new
+  source.  No range needs to be read or generated -- we did not slow this
+  case down.
+* A range that changed in all of base, old source, and old destination, will
+  still need to be read and analyzed.  But the resulting range will be
+  unchanged from the new source, so the merge is trivial: only changes from
+  the old to new destination need be applied, and we expect many ranges not
+  to have such changes.
+
+### Alternative
+
+Rather than immediately retry a merge that loses a race, lakeFS could return
+a failure with an additional "hints" field containing the resulting
+metarange ID.  The client can then try again, supplying lakeFS with the same
+hints.  This allows lakeFS to use the correct metarange in the retry.[^1]
+
+This improves client control of retries.  In some cases it may cause retries
+to be load-balanced onto a less busy lakeFS, which might improve fairness.
+
+### Correctness
+
+Why is it correct to use the new metarange?  It is sufficient to show that
+merging from the new metarange as source will yield the same sequence of
+objects as merging from the old metarange.  How this sequence is split into
+ranges is important for efficiency, but yields an indistinguishable object
+store.
+
+In the sequel it helps to think of "deleted" objects that as having
+particular contents.  The behaviour of deleted objects in a merge regarding
+results and conflicts is exactly the same as of objects with some particular
+unique contents.
+
+| base | src | dst/base' | result/src' | dst' | final result | comment                        |
+|------|-----|-----------|-------------|------|--------------|--------------------------------|
+| A    | B   | A         | B           | A    | B            |                                |
+| A    | A   | B         | B           | B    | B            |                                |
+| A    | A   | A         | A           | B    | B            | (Only) dst changed during race |
+| A    | B   | C         | conflict    | --   | conflict     | Conflicts never proceed        |
+| A    | B   | A         | B           | C    | conflict     | Conflict with updated dst      |
+| A    | A   | B         | B           | C    | C            |                                |
+
+
```

Comment:
I think there is a correction issue when using the conflict resolution flag. The example uses one file and the letter represents a different version of it.

Two concurrent merges

1. Source wins in case of conflict
Base	A
Source B
Dest C
Res B

1. Dest wins in case of conflict
Base A
Source D
Dest C
Res C

The first one finish first so the second one will run with the result metarange - C and the dest wins flag:
 
Base C
Source C
Dest B
Res B

Reply:
I think conflict resolution also works here, actually.  You just need to interpret the new base and source.

Think of this as doing _exactly_ what you would do during manual conflict resolution after losing a race:
* First you merge destination HEAD into your source.  This uses the _reverse_ conflict resolution strategy.
* Now you merge the result back to your destination HEAD.